### PR TITLE
fix: flush of one table might be triggered multiple times

### LINKED ISSUE #183: Flush of one table might be triggered multiple times
### Describe This Problem

Flush is checked on the write path `preprocess_write()`. That method has three conditions to trigger a flush: `should_flush_instance`, `should_flush_space` and `should_flush_table`. When memory usage reached the threshold in the instance or space level a flush is then triggered. Consider two tables belonging to two different `WorkerLocal`s, they can trigger the flush in the instance or space level simultaneously, while these two flushes are likely to operate on the same table because of the logic of selecting tables.

To my understanding, the procedure described above may lead to two problems:
- A table may be flushed multiple times.
- A table may be operated by another `WorkerLocal` which will break our concurrent invariant "The worker is single threaded and holding this is equivalent to holding a write lock"

### Proposal

The key points are
- Can we select and operate tables belonging to other workers?
We can select and schedule a flush for any table, but are only allowed to operate tables belonging to the correct worker.
- Should we ignore memory occupied by memtables that are flushing?
`should_flush_table` will distinguish these memories https://github.com/CeresDB/ceresdb/blob/91553aae4d446b4a3f31aa6e211ae1ffb592576f/analytic_engine/src/table/data.rs#L381 Trigger in instance and space levels can have the same logic. 

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by dust1**
# Which issue does this PR close?

Closes #183 

# Rationale for this change
 
* A table may be flushed multiple times.
* A table may be operated by another `WorkerLocal` which will break our concurrent invariant "The worker is single threaded and holding this is equivalent to holding a write lock"

# What changes are included in this PR?

I added the action to filter out the tables it owns based on the current `workerlocal`.

# Are there any user-facing changes?

# How does this change test

### CODE REVIEW

File: analytic_engine/src/table/data.rs [hunk 1]
```diff
@@ -529,10 +529,16 @@ impl TableDataSet {
         self.table_datas.len()
     }
 
-    /// Find the table which consumes maximum memtable memory usag.
-    pub fn find_maximum_memory_usage_table(&self) -> Option<TableDataRef> {
+    /// Find the table that the current WorkerLocal belongs to and consumes the
+    /// largest memtable memory usage.
+    pub fn find_maximum_memory_usage_table(
```
[the remark this answers is not in the document]
Comment (dust1) [reply analytic_engine/src/table/data.rs:534]: yes. I am wondering about this, does this mean that I need to select the corresponding table by worker_id when I choose to refresh, but do I still need to keep the logic of getting all the tables?

[the remark this answers is not in the document]
Comment (waynexia) [reply analytic_engine/src/table/data.rs:534]: >but do I still need to keep the logic of getting all the tables?

I don't think so. A worker should only operate tables it holds, in any situation.

**Comment by waynexia** [comment]
Thanks @dust1 for this work :heart:

This approach solves the problem cleanly and clearly 👍 And I think we can give one more step to prevent the "invariants" from being broken in other places. How about also checking "worker id" before performing operations like `write` or `flush`?

**Comment by dust1** [comment]
> Thanks @dust1 for this work ❤️
> 
> This approach solves the problem cleanly and clearly 👍 And I think we can give one more step to prevent the "invariants" from being broken in other places. How about also checking "worker id" before performing operations like `write` or `flush`?

of course. Are we going to check the legitimacy of the worker id? Like is it out of bounds?

**Comment by waynexia** [comment]
>Are we going to check the legitimacy of the worker id? Like is it out of bounds?

We won't have too many workers, this ID is not likely to go out of bounds so I think we don't need to check this. For now check whether the `WorkerLocal` and `TableData` are coupled before operations might be enough.

**Comment by dust1** [comment]
> check whether the `WorkerLocal` and `TableData` are coupled before operations might be enough.

ok. i'll add it later🧐

File: analytic_engine/src/instance/write.rs [hunk 2]
```diff
@@ -312,6 +319,16 @@ impl Instance {
             }
         );
 
+        let worker_id = worker_local.worker_id();
+        let worker_num = space.write_group.worker_num();
+        ensure!(
+            table_data.id.as_u64() as usize % worker_num != worker_id,
```
Comment (waynexia) [inline analytic_engine/src/instance/write.rs:325]: The rule might be changed in the future, can we use a function to check this?

**Review by waynexia — COMMENTED** [review]
Not only `write` requires `&mut WorkerLocal`. `flush` (which is the issue originally from) and `dump` also have.

File: analytic_engine/src/instance/write.rs (hunk 2 above)
Comment (dust1) [reply analytic_engine/src/instance/write.rs:325]: ok, i'll fix it tonight

**Review by waynexia — APPROVED** [review]
This change looks good to me. Thanks for fixing it :heart:

**Pull request merged**